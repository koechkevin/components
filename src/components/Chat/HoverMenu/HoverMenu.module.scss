@import "../../../styles/global";

.root {
  align-items: center;
  top: -15px;
  right: 25px;
  background: $color-white;
  border-radius: 4px;
  border: 1px solid $color-border-gray;
  z-index: 999;

  &:global(.ant-row) {
    display: inline-flex;
    position: absolute;

    @include respond-to('sm') {
      display: flex;
      position: static;
    }
  }

  @include respond-to('sm') {
    width: 100%;
    box-shadow: none;
    border-radius: 0;
    border-left-width: 0;
    border-right-width: 0;
  }

  >div {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 32px;
    height: 28px;
    cursor: pointer;

    &:not(:last-child) {
      border-right: 1px solid $color-border-gray;
    }

    @include respond-to('sm') {
      flex: 1;

      &:last-child {
        border-right-width: 0;
      }
    }

    &:hover {
      background: $color-hover-bg;
    }
  }

  svg {
    font-size: 15px;
    color: $color-base-dark;
  }
}

.select {
  background: $color-hover-bg;
}