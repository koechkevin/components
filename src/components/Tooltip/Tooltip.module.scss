@import "../../styles/global";

.tooltip {
  :global(.ant-tooltip-inner) {
    font-size: 12px;
    background-color: $color-base-dark;
  }
}

.light {
  :global {
    .ant-tooltip-inner {
      font-size: 12px;
      color: $color-base-dark;
      background-color: $color-white;
      box-shadow: 0 2px 4px rgba($color: $color-black, $alpha: 0.16);
    }

    .ant-tooltip-arrow::before {
      background-color: $color-white;
    }
  }
}