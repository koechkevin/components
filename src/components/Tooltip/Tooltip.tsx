import { Tooltip as AntdTooltip } from 'antd';
import { AbstractTooltipProps, RenderFunction } from 'antd/lib/tooltip';
import React, { FC, ReactNode } from 'react';

import styles from './Tooltip.module.scss';

interface Props extends AbstractTooltipProps {
  delay?: number;
  light?: boolean;
  title?: RenderFunction | ReactNode;
}

export const Tooltip: FC<Props> = (props) => {
  const { title, light, children, delay, ...restProps } = props;
  const isTouchDevice = () => 'ontouchstart' in window;
  const overlay = light ? styles.light : styles.tooltip;

  return (
    <AntdTooltip
      mouseEnterDelay={delay}
      mouseLeaveDelay={delay}
      title={!isTouchDevice() && title}
      overlayClassName={overlay}
      {...restProps}
    >
      {children}
    </AntdTooltip>
  );
};

Tooltip.defaultProps = {
  delay: 0.0001,
};

export default Tooltip;
