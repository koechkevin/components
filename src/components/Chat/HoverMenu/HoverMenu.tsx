import { faEllipsisH, faLink } from '@fortawesome/pro-light-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Dropdown, Row } from 'antd';
import React, { FC, ReactNode, useState } from 'react';

import { Tooltip } from '../../Tooltip';
import styles from './HoverMenu.module.scss';

interface Props {
  messageId?: string;
  actionMenu: ReactNode;
  startThreadMenu?: ReactNode;
  copyLink?: () => void;
}

const HoverMenu: FC<Props> = (props) => {
  const { actionMenu, copyLink, messageId, startThreadMenu } = props;
  const [visible, setVisible] = useState(false);
  const [dropDownColor, setColor] = useState('');

  const onMore = () => {
    setVisible(true);
    setColor(styles.select);
  };

  return (
    <Row className={styles.root}>
      {startThreadMenu}
      <Tooltip title="Copy link">
        <div onClick={() => messageId && copyLink}>
          <FontAwesomeIcon icon={faLink} />
        </div>
      </Tooltip>
      <Dropdown
        visible={visible}
        trigger={['click']}
        placement="bottomRight"
        overlay={<div onClick={() => setVisible(false)}>{actionMenu}</div>}
      >
        <Tooltip title="More actions">
          <div onClick={onMore} className={dropDownColor}>
            <FontAwesomeIcon icon={faEllipsisH} />
          </div>
        </Tooltip>
      </Dropdown>
    </Row>
  );
};

export default HoverMenu;
