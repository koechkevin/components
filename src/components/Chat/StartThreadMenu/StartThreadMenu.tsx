import { faCommentLines } from '@fortawesome/pro-light-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import React, { FC } from 'react';

import { Tooltip } from '../../Tooltip';

interface Props {
  messageId?: string;
  startThread?: (id: string) => void;
}

const StartThreadMenu: FC<Props> = (props) => {
  const { messageId, startThread } = props;
  return (
    <>
      <Tooltip title="Start a thread" placement="top">
        <div onClick={() => startThread && messageId && startThread(messageId)}>
          <FontAwesomeIcon icon={faCommentLines} />
        </div>
      </Tooltip>
    </>
  );
};

export default StartThreadMenu;
