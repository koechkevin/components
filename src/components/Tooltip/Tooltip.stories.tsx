import { number } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import { Button, Col, Row, Typography } from 'antd';
import React from 'react';

import Tooltip from './Tooltip';

const text = <span>Example text</span>;
const buttonWidth = 70;
const { Title } = Typography;

storiesOf('Tooltip', module).add('default', () => {
  const delay = number('Delay', 0.1);

  return (
    <Row gutter={24}>
      <Col lg={12} md={24} style={{ paddingTop: 60, paddingLeft: 32 }}>
        <Title level={3}>Dark Theme</Title>
        <div style={{ marginLeft: buttonWidth, marginTop: 32, whiteSpace: 'nowrap' }}>
          <Tooltip delay={delay} placement="topLeft" title={text}>
            <Button>TL</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="top" title={text}>
            <Button>Top</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="topRight" title={text}>
            <Button>TR</Button>
          </Tooltip>
        </div>
        <div style={{ width: buttonWidth, float: 'left' }}>
          <Tooltip delay={delay} placement="leftTop" title={text}>
            <Button>LT</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="left" title={text}>
            <Button>Left</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="leftBottom" title={text}>
            <Button>LB</Button>
          </Tooltip>
        </div>
        <div style={{ width: buttonWidth, marginLeft: buttonWidth * 4 + 24 }}>
          <Tooltip delay={delay} placement="rightTop" title={text}>
            <Button>RT</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="right" title={text}>
            <Button>Right</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="rightBottom" title={text}>
            <Button>RB</Button>
          </Tooltip>
        </div>
        <div style={{ marginLeft: buttonWidth, clear: 'both', whiteSpace: 'nowrap' }}>
          <Tooltip delay={delay} placement="bottomLeft" title={text}>
            <Button>BL</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="bottom" title={text}>
            <Button>Bottom</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="bottomRight" title={text}>
            <Button>BR</Button>
          </Tooltip>
        </div>
      </Col>
      <Col lg={12} md={24} style={{ paddingTop: 60, paddingLeft: 32 }}>
        <Title level={3}>White Theme</Title>
        <div style={{ marginLeft: buttonWidth, marginTop: 32, whiteSpace: 'nowrap' }}>
          <Tooltip delay={delay} placement="topLeft" title={text} light>
            <Button>TL</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="top" title={text} light>
            <Button>Top</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="topRight" title={text} light>
            <Button>TR</Button>
          </Tooltip>
        </div>
        <div style={{ width: buttonWidth, float: 'left' }}>
          <Tooltip delay={delay} placement="leftTop" title={text} light>
            <Button>LT</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="left" title={text} light>
            <Button>Left</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="leftBottom" title={text} light>
            <Button>LB</Button>
          </Tooltip>
        </div>
        <div style={{ width: buttonWidth, marginLeft: buttonWidth * 4 + 24 }}>
          <Tooltip delay={delay} placement="rightTop" title={text} light>
            <Button>RT</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="right" title={text} light>
            <Button>Right</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="rightBottom" title={text} light>
            <Button>RB</Button>
          </Tooltip>
        </div>
        <div style={{ marginLeft: buttonWidth, clear: 'both', whiteSpace: 'nowrap' }}>
          <Tooltip delay={delay} placement="bottomLeft" title={text} light>
            <Button>BL</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="bottom" title={text} light>
            <Button>Bottom</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="bottomRight" title={text} light>
            <Button>BR</Button>
          </Tooltip>
        </div>
      </Col>
    </Row>
  );
});
