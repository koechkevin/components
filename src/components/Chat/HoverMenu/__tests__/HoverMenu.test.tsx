import { Dropdown } from 'antd';
import { mount } from 'enzyme';
import React from 'react';

import { Tooltip } from '../../../Tooltip';
import StartThreadMenu from '../../StartThreadMenu/StartThreadMenu';
import HoverMenu from '../HoverMenu';

describe('Tests for Hover menu component', () => {
  const startThreadFn = jest.fn();
  const actionMenu = <div>Mock Component</div>;
  const startThreadMenu = <StartThreadMenu startThread={startThreadFn} messageId={'sample message id'} />;

  const props = {
    actionMenu,
    startThreadMenu,
  };

  const wrapper = mount(<HoverMenu {...props} />);

  it('should mount hover menu component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should simulate hover on the tooltip', () => {
    const toolTipComponent = wrapper.find(Tooltip).first();

    expect(toolTipComponent.simulate('mouseOver').contains('Start a thread'));
  });

  it('should simulate clicking of start thread tooltip', () => {
    const toolTipComponent = wrapper.find(Tooltip).first();
    const clickableDivElement = toolTipComponent.find('div').first();

    clickableDivElement.simulate('click');

    expect(startThreadFn).toHaveBeenCalled();
  });

  it('should simulate rendering of drop down to the bottom right', () => {
    const dropDown = wrapper.find(Dropdown).first();

    expect(dropDown.props().placement).toEqual('bottomRight');
  });

  it('should simulate clicking of the bottom right drop down', () => {
    const dropDown = wrapper.find(Dropdown).first();
    dropDown
      .find('div')
      .first()
      .simulate('click');

    expect(wrapper.props().actionMenu).toEqual(actionMenu);
  });

  it('should render more actions tooltip', () => {
    const toolTip = wrapper.find('Tooltip').last();

    expect(toolTip.props().title).toEqual('More actions');
  });

  it('should simulate click of more actions tooltip to set color', () => {
    const toolTip = wrapper
      .find('Tooltip')
      .last()
      .find('div')
      .first()
      .simulate('click');

    expect(toolTip.find('.select').exists()).toEqual(true);
  });
});
