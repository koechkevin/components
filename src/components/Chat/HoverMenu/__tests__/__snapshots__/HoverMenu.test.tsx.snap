// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Hover menu component should mount hover menu component 1`] = `ReactWrapper {}`;
